src/Stratis.Features.FederatedPeg/CollateralChecker.cs

```diff
@@ -0,0 +1,195 @@
+        private const int RequiredConfirmations = 1;
+
+        private const int CollateralUpdateIntervalSeconds = 20;
+
+        /// <summary>Deposits mapped by federation member.</summary>
+        /// <remarks>
+        /// Deposits are not updated if federation member doesn't have collateral requirement enabled.
+        /// All access should be protected by <see cref="locker"/>.
+        /// </remarks>
+        private Dictionary<string, Money> depositsByAddress;
+
+        private Task updateCollateralContinuouslyTask;
+
+        public CollateralChecker(ILoggerFactory loggerFactory, IHttpClientFactory httpClientFactory, FederationGatewaySettings settings,
+            IFederationManager federationManager, ISignals signals)
+        {
+            this.federationManager = federationManager;
+            this.signals = signals;
+
+            this.cancellationSource = new CancellationTokenSource();
+            this.locker = new object();
+            this.depositsByAddress = new Dictionary<string, Money>();
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.blockStoreClient = new BlockStoreClient(loggerFactory, httpClientFactory, settings.CounterChainApiPort);
+        }
+
+        public async Task InitializeAsync()
+        {
+            this.memberAddedToken = this.signals.Subscribe<FedMemberAdded>(this.OnFedMemberAdded);
+            this.memberKickedToken = this.signals.Subscribe<FedMemberKicked>(this.OnFedMemberKicked);
+
+            foreach (CollateralFederationMember federationMember in this.federationManager.GetFederationMembers().Cast<CollateralFederationMember>().Where(x => x.CollateralAmount != null && x.CollateralAmount > 0))
+                this.depositsByAddress.Add(federationMember.CollateralMainchainAddress, null);
+
+            while (true)
+            {
+                bool success = await this.UpdateCollateralInfoAsync(this.cancellationSource.Token).ConfigureAwait(false);
+
+                this.logger.LogWarning("Failed to update collateral. Ensure that mainnet gateway node is running and API is enabled. " +
+                                       "Node will not continue initialization before another gateway node responds.");
+
+                if (!success)
+                {
+                    try
+                    {
+                        await Task.Delay(3000, this.cancellationSource.Token).ConfigureAwait(false);
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        this.logger.LogTrace("(-)[CANCELLED]");
+                        return;
+                    }
+                }
+                else
+                {
+                    break;
+                }
+            }
+
+            this.updateCollateralContinuouslyTask = this.UpdateCollateralInfoContinuouslyAsync();
+        }
+
+        private async Task UpdateCollateralInfoContinuouslyAsync()
```

Comment:
Put this in AsyncProvider?

Reply:
What do you mean? AsyncProvider is a kind of a factory for AsyncQueue and AsyncLoop. I'm not using those in this class.

Reply:
This is a perfect example of using an AsyncLoop.
But we will add an additional method to the AsyncProvider to return a tracked `Task` so it will effectively be tracked @MithrilMan is aware.

Reply:
Ok, sounds good. 
Tracking tasks is a good idea; I primarily use them over the async loop.